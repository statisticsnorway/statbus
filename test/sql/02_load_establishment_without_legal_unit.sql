BEGIN;

\echo "Setting up Statbus to load establishments without legal units"

\echo "User selected the Activity Category Standard"
INSERT INTO settings(activity_category_standard_id,only_one_setting)
SELECT id, true FROM activity_category_standard WHERE code = 'nace_v2.1'
ON CONFLICT (only_one_setting)
DO UPDATE SET
   activity_category_standard_id =(SELECT id FROM activity_category_standard WHERE code = 'nace_v2.1')
   WHERE settings.id = EXCLUDED.id;
;
SELECT activity_category_standard_id FROM public.settings;

\echo "User uploads the sample activity categories"
\copy public.activity_category_available_custom(path,name,description) FROM 'app/public/activity_category_norway.csv' WITH (FORMAT csv, DELIMITER ',', QUOTE '"', HEADER true);
SELECT count(*) FROM public.activity_category_available;

\echo "User uploads the sample regions"
\copy public.region_upload(path, name) FROM 'app/public/norway-regions-2024.csv' WITH (FORMAT csv, DELIMITER ',', QUOTE '"', HEADER true);
SELECT count(*) FROM public.region;

\echo "User uploads the sample legal forms"
\copy public.legal_form_custom_only(code,name) FROM 'app/public/legal_form_norway.csv' WITH (FORMAT csv, DELIMITER ',', QUOTE '"', HEADER true);
SELECT count(*) FROM public.legal_form_available;

\echo "User uploads the sample sectors"
\copy public.sector_custom_only(path,name,description) FROM 'app/public/sector_norway.csv' WITH (FORMAT csv, DELIMITER ',', QUOTE '"', HEADER true);
SELECT count(*) FROM public.sector_available;

SELECT
    (SELECT COUNT(*) FROM public.establishment) AS establishment_count,
    (SELECT COUNT(*) FROM public.legal_unit) AS legal_unit_count,
    (SELECT COUNT(*) FROM public.enterprise) AS enterprise_count;
\echo "User uploads establishments without legal_unit"
\copy public.import_establishment_current_without_legal_unit(tax_ident,name,birth_date,death_date,physical_address_part1,physical_postal_code,physical_postal_place,physical_region_code,physical_country_iso_2,postal_address_part1,postal_postal_code,postal_postal_place,postal_region_code,postal_country_iso_2,primary_activity_category_code,secondary_activity_category_code,employees) FROM 'test/data/norwegian-establishments-without-legal-unit.csv' WITH (FORMAT csv, DELIMITER ',', QUOTE '"', HEADER true);
SELECT
    (SELECT COUNT(*) FROM public.establishment) AS establishment_count,
    (SELECT COUNT(*) FROM public.legal_unit) AS legal_unit_count,
    (SELECT COUNT(*) FROM public.enterprise) AS enterprise_count;

\echo "Refreshing materialized views"
-- Exclude the refresh_time_ms as it will vary.
SELECT view_name FROM statistical_unit_refresh_now();

\x
SELECT unit_type, name, external_idents, stats, jsonb_pretty(stats_summary) AS stats_summary
FROM statistical_unit ORDER BY name, unit_type;

\echo "Checking statistics"
SELECT unit_type
     , count(*)
     , jsonb_pretty(jsonb_agg(DISTINCT invalid_codes) FILTER (WHERE invalid_codes IS NOT NULL)) AS invalid_codes
     , jsonb_pretty(jsonb_stats_summary_merge_agg(stats_summary)) AS stats_summary
 FROM statistical_unit
 GROUP BY unit_type;
\x

ROLLBACK;
