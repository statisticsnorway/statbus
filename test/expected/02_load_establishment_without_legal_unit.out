BEGIN;
\echo "Setting up Statbus to load establishments without legal units"
"Setting up Statbus to load establishments without legal units"
\echo "User selected the Activity Category Standard"
"User selected the Activity Category Standard"
INSERT INTO settings(activity_category_standard_id,only_one_setting)
SELECT id, true FROM activity_category_standard WHERE code = 'nace_v2.1'
ON CONFLICT (only_one_setting)
DO UPDATE SET
   activity_category_standard_id =(SELECT id FROM activity_category_standard WHERE code = 'nace_v2.1')
   WHERE settings.id = EXCLUDED.id;
;
SELECT acs.code
  FROM public.settings AS s
  JOIN activity_category_standard AS acs
    ON s.activity_category_standard_id = acs.id;
   code    
-----------
 nace_v2.1
(1 row)

\echo "User uploads the sample activity categories"
"User uploads the sample activity categories"
\copy public.activity_category_available_custom(path,name,description) FROM 'samples/norway/activity_category/activity_category_norway.csv' WITH (FORMAT csv, DELIMITER ',', QUOTE '"', HEADER true);
SELECT count(*) FROM public.activity_category_available;
 count 
-------
  2215
(1 row)

\echo "User uploads the sample regions"
"User uploads the sample regions"
\copy public.region_upload(path, name) FROM 'samples/norway/regions/norway-regions-2024.csv' WITH (FORMAT csv, DELIMITER ',', QUOTE '"', HEADER true);
SELECT count(*) FROM public.region;
 count 
-------
   417
(1 row)

\echo "User uploads the sample legal forms"
"User uploads the sample legal forms"
\copy public.legal_form_custom_only(code,name) FROM 'samples/norway/legal_form/legal_form_norway.csv' WITH (FORMAT csv, DELIMITER ',', QUOTE '"', HEADER true);
SELECT count(*) FROM public.legal_form_available;
 count 
-------
    46
(1 row)

\echo "User uploads the sample sectors"
"User uploads the sample sectors"
\copy public.sector_custom_only(path,name,description) FROM 'samples/norway/sector/sector_norway.csv' WITH (FORMAT csv, DELIMITER ',', QUOTE '"', HEADER true);
SELECT count(*) FROM public.sector_available;
 count 
-------
    33
(1 row)

SELECT code, name, active, custom FROM public.data_source_available;
 code |               name               | active | custom 
------+----------------------------------+--------+--------
 me   | Manually entered                 | t      | f
 nlr  | National Legal Business Registry | t      | f
 ntr  | National Tax Registry            | t      | f
(3 rows)

\echo "User uploads the sample data sources"
"User uploads the sample data sources"
\copy public.data_source_custom(code,name) FROM 'test/data/02_norwegian_data_source.csv' WITH (FORMAT csv, DELIMITER ',', QUOTE '"', HEADER true);
SELECT code, name, active, custom FROM public.data_source_available;
 code  |               name               | active | custom 
-------+----------------------------------+--------+--------
 brreg | Brønnøysundregistrene            | t      | t
 mi    | Manuell inntasting               | t      | t
 nav   | Arbeids- og velferdsforvaltninga | t      | t
(3 rows)

SELECT count(*) FROM public.data_source_available;
 count 
-------
     3
(1 row)

SELECT
    (SELECT COUNT(DISTINCT id) AS distinct_unit_count FROM public.establishment) AS establishment_count,
    (SELECT COUNT(DISTINCT id) AS distinct_unit_count FROM public.legal_unit) AS legal_unit_count,
    (SELECT COUNT(DISTINCT id) AS distinct_unit_count FROM public.enterprise) AS enterprise_count;
 establishment_count | legal_unit_count | enterprise_count 
---------------------+------------------+------------------
                   0 |                0 |                0
(1 row)

\echo "User uploads establishments without legal_unit"
"User uploads establishments without legal_unit"
\copy public.import_establishment_current_without_legal_unit(tax_ident,name,birth_date,death_date,physical_address_part1,physical_postal_code,physical_postal_place,physical_region_code,physical_country_iso_2,postal_address_part1,postal_postal_code,postal_postal_place,postal_region_code,postal_country_iso_2,primary_activity_category_code,secondary_activity_category_code,employees,data_source_code) FROM 'test/data/02_norwegian-establishments-without-legal-unit.csv' WITH (FORMAT csv, DELIMITER ',', QUOTE '"', HEADER true);
SELECT
    (SELECT COUNT(DISTINCT id) AS distinct_unit_count FROM public.establishment) AS establishment_count,
    (SELECT COUNT(DISTINCT id) AS distinct_unit_count FROM public.legal_unit) AS legal_unit_count,
    (SELECT COUNT(DISTINCT id) AS distinct_unit_count FROM public.enterprise) AS enterprise_count;
 establishment_count | legal_unit_count | enterprise_count 
---------------------+------------------+------------------
                   5 |                0 |                5
(1 row)

\echo "Refreshing materialized views"
"Refreshing materialized views"
-- Exclude the refresh_time_ms as it will vary.
SELECT view_name FROM statistical_unit_refresh_now();
         view_name         
---------------------------
 statistical_unit
 activity_category_used
 region_used
 sector_used
 data_source_used
 legal_form_used
 country_used
 statistical_unit_facet
 statistical_history
 statistical_history_facet
(10 rows)

\x
SELECT unit_type, name, external_idents, data_source_codes, stats, jsonb_pretty(stats_summary) AS stats_summary
FROM statistical_unit ORDER BY name, unit_type;
-[ RECORD 1 ]-----+---------------------------------------------
unit_type         | establishment
name              | GRAPHICS BY MIA
external_idents   | {"tax_ident": "825481982"}
data_source_codes | {brreg}
stats             | {}
stats_summary     | {                                           +
                  | }
-[ RECORD 2 ]-----+---------------------------------------------
unit_type         | enterprise
name              | GRAPHICS BY MIA
external_idents   | {"tax_ident": "825481982"}
data_source_codes | {brreg}
stats             | 
stats_summary     | {                                           +
                  | }
-[ RECORD 3 ]-----+---------------------------------------------
unit_type         | establishment
name              | KRAN & SPESIALLØFT AS
external_idents   | {"tax_ident": "895406732"}
data_source_codes | {brreg}
stats             | {"employees": 19}
stats_summary     | {                                           +
                  |     "employees": {                          +
                  |         "max": 19,                          +
                  |         "min": 19,                          +
                  |         "sum": 19,                          +
                  |         "mean": 19,                         +
                  |         "type": "number",                   +
                  |         "count": 1,                         +
                  |         "stddev": 0,                        +
                  |         "variance": 0,                      +
                  |         "sum_sq_diff": 0,                   +
                  |         "coefficient_of_variation_pct": 0   +
                  |     }                                       +
                  | }
-[ RECORD 4 ]-----+---------------------------------------------
unit_type         | enterprise
name              | KRAN & SPESIALLØFT AS
external_idents   | {"tax_ident": "895406732"}
data_source_codes | {brreg}
stats             | 
stats_summary     | {                                           +
                  |     "employees": {                          +
                  |         "max": 19,                          +
                  |         "min": 19,                          +
                  |         "sum": 19,                          +
                  |         "mean": 19.00,                      +
                  |         "type": "number",                   +
                  |         "count": 1,                         +
                  |         "stddev": 0.00,                     +
                  |         "variance": 0.00,                   +
                  |         "sum_sq_diff": 0.00,                +
                  |         "coefficient_of_variation_pct": 0.00+
                  |     }                                       +
                  | }
-[ RECORD 5 ]-----+---------------------------------------------
unit_type         | establishment
name              | NETT & EL TEKNIKK AS
external_idents   | {"tax_ident": "819097232"}
data_source_codes | {brreg}
stats             | {}
stats_summary     | {                                           +
                  | }
-[ RECORD 6 ]-----+---------------------------------------------
unit_type         | enterprise
name              | NETT & EL TEKNIKK AS
external_idents   | {"tax_ident": "819097232"}
data_source_codes | {brreg}
stats             | 
stats_summary     | {                                           +
                  | }
-[ RECORD 7 ]-----+---------------------------------------------
unit_type         | establishment
name              | NORSVIN SA TESTESTASJON DELTA
external_idents   | {"tax_ident": "873124032"}
data_source_codes | {brreg}
stats             | {"employees": 16}
stats_summary     | {                                           +
                  |     "employees": {                          +
                  |         "max": 16,                          +
                  |         "min": 16,                          +
                  |         "sum": 16,                          +
                  |         "mean": 16,                         +
                  |         "type": "number",                   +
                  |         "count": 1,                         +
                  |         "stddev": 0,                        +
                  |         "variance": 0,                      +
                  |         "sum_sq_diff": 0,                   +
                  |         "coefficient_of_variation_pct": 0   +
                  |     }                                       +
                  | }
-[ RECORD 8 ]-----+---------------------------------------------
unit_type         | enterprise
name              | NORSVIN SA TESTESTASJON DELTA
external_idents   | {"tax_ident": "873124032"}
data_source_codes | {brreg}
stats             | 
stats_summary     | {                                           +
                  |     "employees": {                          +
                  |         "max": 16,                          +
                  |         "min": 16,                          +
                  |         "sum": 16,                          +
                  |         "mean": 16.00,                      +
                  |         "type": "number",                   +
                  |         "count": 1,                         +
                  |         "stddev": 0.00,                     +
                  |         "variance": 0.00,                   +
                  |         "sum_sq_diff": 0.00,                +
                  |         "coefficient_of_variation_pct": 0.00+
                  |     }                                       +
                  | }
-[ RECORD 9 ]-----+---------------------------------------------
unit_type         | establishment
name              | VIKEDAL LEGEKONTOR
external_idents   | {"tax_ident": "911856174"}
data_source_codes | {brreg}
stats             | {"employees": 5}
stats_summary     | {                                           +
                  |     "employees": {                          +
                  |         "max": 5,                           +
                  |         "min": 5,                           +
                  |         "sum": 5,                           +
                  |         "mean": 5,                          +
                  |         "type": "number",                   +
                  |         "count": 1,                         +
                  |         "stddev": 0,                        +
                  |         "variance": 0,                      +
                  |         "sum_sq_diff": 0,                   +
                  |         "coefficient_of_variation_pct": 0   +
                  |     }                                       +
                  | }
-[ RECORD 10 ]----+---------------------------------------------
unit_type         | enterprise
name              | VIKEDAL LEGEKONTOR
external_idents   | {"tax_ident": "911856174"}
data_source_codes | {brreg}
stats             | 
stats_summary     | {                                           +
                  |     "employees": {                          +
                  |         "max": 5,                           +
                  |         "min": 5,                           +
                  |         "sum": 5,                           +
                  |         "mean": 5.00,                       +
                  |         "type": "number",                   +
                  |         "count": 1,                         +
                  |         "stddev": 0.00,                     +
                  |         "variance": 0.00,                   +
                  |         "sum_sq_diff": 0.00,                +
                  |         "coefficient_of_variation_pct": 0.00+
                  |     }                                       +
                  | }

\echo "Checking statistics"
"Checking statistics"
SELECT unit_type
     , COUNT(DISTINCT unit_id) AS distinct_unit_count
     , jsonb_pretty(jsonb_agg(DISTINCT invalid_codes) FILTER (WHERE invalid_codes IS NOT NULL)) AS invalid_codes
     , jsonb_pretty(jsonb_stats_summary_merge_agg(stats_summary)) AS stats_summary
 FROM statistical_unit
 GROUP BY unit_type;
-[ RECORD 1 ]-------+----------------------------------------------
unit_type           | establishment
distinct_unit_count | 5
invalid_codes       | 
stats_summary       | {                                            +
                    |     "employees": {                           +
                    |         "max": 19,                           +
                    |         "min": 5,                            +
                    |         "sum": 40,                           +
                    |         "mean": 13.33,                       +
                    |         "type": "number",                    +
                    |         "count": 3,                          +
                    |         "stddev": 7.37,                      +
                    |         "variance": 54.33,                   +
                    |         "sum_sq_diff": 108.67,               +
                    |         "coefficient_of_variation_pct": 55.28+
                    |     }                                        +
                    | }
-[ RECORD 2 ]-------+----------------------------------------------
unit_type           | enterprise
distinct_unit_count | 5
invalid_codes       | 
stats_summary       | {                                            +
                    |     "employees": {                           +
                    |         "max": 19,                           +
                    |         "min": 5,                            +
                    |         "sum": 40,                           +
                    |         "mean": 13.33,                       +
                    |         "type": "number",                    +
                    |         "count": 3,                          +
                    |         "stddev": 7.37,                      +
                    |         "variance": 54.33,                   +
                    |         "sum_sq_diff": 108.67,               +
                    |         "coefficient_of_variation_pct": 55.28+
                    |     }                                        +
                    | }

\x
\i test/rollback_unless_persist_is_specified.sql
---------------------------------------------------------------------------
-- Support development loading of the data without rollback using
--   ./devops/manage-statbus.sh psql --variable=PERSIST=true < test/sql/01_load_web_examples.sql
-- Ref. https://stackoverflow.com/a/32597876/1023558
\set PERSIST :PERSIST
-- now PERSIST is set to the string ':PERSIST' if was not already set.
-- Checking it using a CASE statement:
SELECT CASE
  WHEN :'PERSIST'= ':PERSIST'
  THEN 'false'
  ELSE :'PERSIST'
END::BOOL AS "PERSIST" \gset
-- < \gset call at end of the query to set variable.
\if :PERSIST
COMMIT;
\else
ROLLBACK;
\endif
